3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id_1), "This cat 1 does not exist");
+            ensure!(<Kitties<T>>::exists(kitty_id_2), "This cat 2 does not exist");
+
+            let nonce = <Nonce<T>>::get();
+            let random_hash = (<system::Module<T>>::random_seed(), &sender, nonce)
+                .using_encoded(<T as system::Trait>::Hashing::hash);
+
+            let kitty_1 = Self::kitty(kitty_id_1);
+            let kitty_2 = Self::kitty(kitty_id_2);
+
+            let mut final_dna = kitty_1.dna;
+
+            for (i, (dna_2_element, r)) in kitty_2.dna.as_ref().iter().zip(random_hash.as_ref().iter()).enumerate() {
+                if r % 2 == 0 {
+                    final_dna.as_mut()[i] = *dna_2_element;
+                }
+            }
+
+            let new_kitty = Kitty {
+                id: random_hash,
+                dna: final_dna,
+                price: <T::Balance as As<u64>>::sa(0),
+                gen: cmp::max(kitty_1.gen, kitty_2.gen) + 1,
+            };
+
+            Self::_mint(sender, random_hash, new_kitty)?;
+
+            <Nonce<T>>::mutate(|n| *n += 1);
+
+            Ok(())
+        }
+
+        fn create_auction(origin, kitty_id: T::Hash, min_bid: T::Balance, expiry_in_minutes: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner == sender, "You can't set an auction for a cat you don't own");
+
+            let expiry = <system::Module<T>>::block_number() + T::BlockNumber::sa(expiry_in_minutes*12);
+            ensure!(expiry <= <system::Module<T>>::block_number() + Self::auction_duration_limit(), "The auction duration should be less than 5 days");
+
+            let new_auction = Auction {
+                kitty_id,
+                kitty_owner: owner,
+                expiry,
+                min_bid,
+                high_bid: min_bid,
+                high_bidder: sender,
+            };
+
+            <KittyAuction<T>>::insert(kitty_id, &new_auction);
+            <Auctions<T>>::insert(expiry, vec![new_auction.clone()]);
+
+            Self::deposit_event(RawEvent::Auction(kitty_id, min_bid, expiry));
+
+            Ok (())
+        }
+
+        fn bid_auction(origin, kitty_id: T::Hash, bid: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't bid for your own cat");
+
+            let mut auction = Self::auction_of(kitty_id).ok_or("No auction for this cat")?;
+            ensure!(<system::Module<T>>::block_number() < auction.expiry, "This auction is expired.");
```

Comment:
It looks like this line handles a specific corner case where a user bids exactly on the block where the auction expires. If `block_number` is greater than `auction.expiry` then `on_finalise` should have already removed it.

I think this is not a bad check to have in case your storage gets messed up, but I think it should at least be consistant with `on_finalise` and check that `block_number() <= auction.expiry`